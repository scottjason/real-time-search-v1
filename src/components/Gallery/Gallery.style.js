import styled from 'styled-components';

export const Container = styled.div`
  position: relative;
  max-width: 100%;
  background-color: black;
  z-index: 1;
`;
export const GridWrap = styled.div`
   position: relative;
   padding: 30px;
   width: auto;
   height: auto;
`;

export const Grid = styled.div`
  margin: 120px auto 0 auto;
  display: grid;
  grid-template-columns: 33% 33% 33%;
  grid-gap: 10px;
`;

export const NoResults = styled.div`
  color: #00bbc4;
  margin-top: 5px;
  font-size: 1.2rem;
  text-align: center;
`;

export const Navbar = styled.nav`
  position: fixed;
  top: 0;
  width: 100%;
  height: auto;
  padding-top: 60px;
  padding-bottom: 15px;
  background-color: black;
  z-index: 2;
`;