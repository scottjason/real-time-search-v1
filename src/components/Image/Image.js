import React from 'react';
import {Img} from './Image.style';
import LazyLoad from 'react-lazyload';

const Image = ({ item }) => {
  return(
    <LazyLoad height={300}>
      <Img imageUrl={item.imageUrl} />
    </LazyLoad>
  )
}

export default Image;
