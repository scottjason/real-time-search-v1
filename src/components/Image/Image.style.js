import styled from 'styled-components';
import media from 'styled-media-query';
const hdUrl = 'https://images.unsplash.com/photo-1478562853135-c3c9e3ef7905';

export const Img = styled.div`
  min-height: 300px;
  background-size: cover;
  background-repeat: no-repeat;
  background-postion: center;
  background-image: url(${props => props.imageUrl});
  border: 1px solid rgba(225, 225, 225, .4);
  transition: all ease 150ms;
  box-sizing: border-box;
  ${media.lessThan('medium')`
    /* screen width is less than 768px (medium) */
    min-height: 100px;
  `}
`;
