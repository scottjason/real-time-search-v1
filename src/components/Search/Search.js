import React from 'react';
import {Input, Form, Container} from './Search.style';
import Pagination from '../Pagination/Pagination';
import Stats from '../Stats/Stats';
import debounce from 'debounce';

const Search = props => {

  let inputRef = React.createRef();
  
  const getInputVal = () => {
    return inputRef.current ? inputRef.current.value : '';
  }

  const debouncedSearch = debounce(()=> {
    props.performSearch(getInputVal(), 1);
  }, 100);
  
  const onSubmit = e => {
    e && e.preventDefault();
    props.performSearch(getInputVal(), 1);
  }

  const fetchPrev = () => props.fetchPrev(getInputVal());
  const fetchNext = () => props.fetchNext(getInputVal());
  
  const enablePrev = props.paginationOpts && props.paginationOpts.left;
  const enableNext = props.paginationOpts && props.paginationOpts.right;

  return(
    <Container>
      <Stats stats={props.stats} />
      <Form onSubmit={onSubmit}>
        <Pagination
          enabled={enablePrev}
          term={getInputVal()}
          fetchPrev={fetchPrev}
        />
        <Input
          autoFocus={true}
          placeholder={'Search images...'}
          defaultValue={'nature'}
          onChange={debouncedSearch}
          ref={inputRef}
        />
        <Pagination
          enabled={enableNext}
          term={getInputVal()}
          fetchNext={fetchNext}
        />
      </Form>
    </Container>
  )
}

export default Search;