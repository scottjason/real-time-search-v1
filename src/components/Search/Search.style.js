import styled from 'styled-components';
import media from 'styled-media-query';

export const Input = styled.input`
  width: 600px;
  height: 50px;
  background-color: #333;
  outline: none;
  border: 1px solid rgba(0, 0, 0, .2);
  padding: 10px 55px 10px 20px;
  font-size: 1.4rem;
  color: white;
  font-weight: 300;
  ${media.lessThan('medium')`
    /* screen width is less than 768px (medium) */
    height: 30px;
    margin: 0 auto;
  `}
`;

export const Form = styled.form`
  position: relative;
  width: 600px;
  margin: 0 auto;
  display: flex;
  justify-content: center;
  ${media.lessThan('medium')`
    /* screen width is less than 768px (medium) */
    width: 95%;
  `}
`;

