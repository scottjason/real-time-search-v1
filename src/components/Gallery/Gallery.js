import React from 'react';
import Image from '../Image/Image';
import Search from '../Search/Search';
import { Grid, NoResults, Container } from './Gallery.style';

const Gallery = props => {
  const getMsg = () => {
    return props.isRequesting ? 'Loading...' : 'No Results';
  }
  return(
    <Container>
      <Search
        isRequesting={props.isRequesting}
        stats={props.stats}
        fetchPrev={props.fetchPrev}
        fetchNext={props.fetchNext}
        performSearch={props.performSearch}
        paginationOpts={props.paginationOpts}
      />
      {!props.isRequesting &&
        <Grid>
          {props.images.map((item, i)=> {
            return(
              <Image
                key={item.id}
                item={item}
              />
            )
          })}
        </Grid> 
      }
      {!props.images.length &&
        <NoResults>
          {getMsg()}
        </NoResults>
      }
    </Container>
  )
}


export default Gallery;
