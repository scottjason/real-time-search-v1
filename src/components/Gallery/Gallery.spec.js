import React from 'react';
import { shallow, mount } from 'enzyme';
import Gallery from './Gallery';
import Search from '../Search/Search';
import Image from '../Image/Image';

/**
 *
 * Unit Testing <Gallery/>
 *
*/

describe('Gallery', () => {

  let wrapper;
  let mockImages = [
    {id: 0, imageUrl: 'https://picsum.photos/500'},
    {id: 1, imageUrl: 'https://picsum.photos/500'},
    {id: 2, imageUrl: 'https://picsum.photos/500'},
  ];

  beforeAll(() => {
    let childA = shallow(<Search />)
    let childB = shallow(<Image item={mockImages[0]} />)
    wrapper = mount(
       <Gallery images={mockImages}>
         {childA}
         {childB}
       </Gallery>
      )
  });

  /**
  * High-Level Checks
  */
  test('it should exist', () => {
    expect(wrapper).toExist(true);
  });

  test('it should have correct name', () => {
    expect(wrapper.name()).toEqual('Gallery');
  });

  test('it should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  test('it should should have two children', () => {
    expect(wrapper.props().children.length).toEqual(2)
  }); 
  
  /**
  * Props
  */
  test('it should have images defined', () => {
    expect(wrapper.props().images).toBeDefined();
  });
  
  test('it should have an images array', () => {
    expect(wrapper.props().images).toBeInstanceOf(Array)
  });
  
  test('it should have three images', () => {
    expect(wrapper.props().images.length).toEqual(3)
  });

  /**
  * Search Component
  */
  test('it should should contain a <Search /> component', () => {
    expect(wrapper.contains(<Search />)).toEqual(true)
  });  
  
  test('it should should contain a form element', () => {
    expect(wrapper.find('form')).toExist(true)
  });
  
  test('it should should contain an input element', () => {
    expect(wrapper.find('input')).toExist(true)
  });

  test('search input should have right placeholder', () => {
    expect(wrapper.find('input').props().placeholder).toEqual('Search images...')
  });

});