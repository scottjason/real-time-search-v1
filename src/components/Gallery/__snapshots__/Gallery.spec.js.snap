// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery it should match snapshot 1`] = `undefined`;
